File: DivideInteger.java
// Time Complexity : O(log n) where n is the divisor
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Understanding Shifts concept
/* Your code here along with comments explaining your approach: Do arithematic shifts till dividend >= divisor. Increment the number of times, you
are doing arithematic shifts. If the number exceeds dividend, hold back and  check if the current dividend is greater than the divisor. If yes,
repeat the same process again till you are finish with a scenario where divisor exceeds the dividend. If both the numbers are opposite, return
-result else return result only.
*/
class Solution {
    public int divide(int dividend, int divisor) {
        if(divisor == 0 || dividend == 0) return 0;
        if(divisor == -1 && dividend == Integer.MIN_VALUE)                                                          // Edge case
            return Integer.MAX_VALUE;
        long dividendL = Math.abs((long)dividend);
        long divisorL = Math.abs((long)divisor);                                                            // Make both the numbers positive
        int result = 0;
        while(dividendL >= divisorL){
            int shifts =0;
            while(dividendL >= (divisorL << shifts)){                                                       // Whilte dividend > divisor, do arithmatic shifts
                shifts++;                                                                               // Count the number of shifts = quotient
            }
           shifts--;                                                                                    // Overshoot
            dividendL =dividendL - (divisorL << shifts);                                            // Still dividend is greater than divisor
            result += 1 << shifts;
        }
        if((dividend > 0 && divisor > 0) || (dividend < 0 && divisor < 0)) return result;                           // Both have same sign
        else return -result;                                                                            // Both are of different sign
    }
}

File: DivideInteger.py
#All TC passed on leetcode


class Solution:
    def divide(self, dividend: int, divisor: int) -> int:

        #Here we keep left shifting divisor until it exceeds dividend. Then we check if diff of dividend and 
        #left shifted divisor is greater than divisor. if yes we update dividend an repeat above steps. 
        #The number of times we left shift raised to 2 is the result.
        #Time complexity - O(k) - where k<quotient<dividend
        #Space complexity - O(1)
        cDivisor = abs(divisor)
        cDividend = abs(dividend)
        res = 0

        if dividend == -2147483648 and divisor==-1:
            return 2147483647

        while cDividend >= cDivisor:
            shifts = 0

            while cDividend >= (cDivisor << shifts):
                shifts += 1
            
            
            shifts -= 1
            cDividend = cDividend - (cDivisor << shifts)
            res += pow(2, shifts)

        if divisor>0 and dividend>0 or divisor<0 and dividend<0:
            return res

        return -res



#------------------------------------OR------------------------------------------

        #Here we keep adding divisor until it exceeds dividend. The number of times we add it is the result.
        #Time complexity - O(quotient) - sometime if divisor is 1 then its O(n) resulting in time exceed 
        #Space complexity - O(1)
        count = 0
        divisorSum = 0
        cDivisor = abs(divisor)
        cDividend = abs(dividend)
        while divisorSum<=cDividend:
            divisorSum = divisorSum + cDivisor
            count+=1
        if divisor>0 and dividend>0 or divisor<0 and dividend<0:
            return count-1
        
        return -(count-1)
            